File: 2559-count-vowel-strings-in-ranges/main.cpp
class Solution {
public:
    vector<int> vowelStrings(vector<string>& words, vector<vector<int>>& queries) {
        int n = words.size();
        vector<int> prefix_sum(n);
        int _sum = 0;

        for (int i = 0; i < n; i++) {
            string curr_word = words[i];
            if (isVowel(curr_word[0]) && isVowel(curr_word.back())) {
                _sum++;
            }
            prefix_sum[i] = _sum;
        }

        n = queries.size();
        vector<int> ans(n);
        for (int i = 0; i < n; i++ ){
            vector<int> curr_query = queries[i];
            ans[i] = prefix_sum[curr_query[1]] - (curr_query[0] == 0 ? 0 : prefix_sum[curr_query[0] - 1]);
        }

        return ans;
    }

private:
    bool isVowel(char letter) {
        static const unordered_set<char> vowels{'a', 'e', 'i', 'o', 'u'};
        return vowels.count(tolower(letter)) > 0;    
    }
};

File: 2559-count-vowel-strings-in-ranges/main.py
class Solution:
    def vowelStrings(self, words: List[str], queries: List[List[int]]) -> List[int]:
        
        vowels = set(['a', 'e', 'i', 'o', 'u'])
        n = len(words)


        prefix_sum = [0] * n
        _sum = 0
        for i in range(n):
            curr_word = words[i]
            if curr_word[0] in vowels and curr_word[-1] in vowels:
                _sum += 1
            
            prefix_sum[i] = _sum
 

        m = len(queries)
        ans = [0] * m
        for i in range(m):
            curr_query = queries[i]
            ans[i] = prefix_sum[curr_query[1]] - (0 if curr_query[0] == 0 else prefix_sum[curr_query[0] - 1])

        return ans
